File: heap_data_structure.java
import java.io.File;
import java.util.*;

class Barang implements Comparable<Barang> {
    int id;
    String nama;

    public Barang(int id, String nama) {
        this.id = id;
        this.nama = nama;
    }

    @Override
    public int compareTo(Barang other) {
        return Integer.compare(this.id, other.id);
    }

    @Override
    public String toString() {
        return String.format("ID: %-5d | Nama: %s", id, nama);
    }
}

public class heap_data_structure {
    private PriorityQueue<Barang> minHeap = new PriorityQueue<>();
    private PriorityQueue<Barang> maxHeap = new PriorityQueue<>(Collections.reverseOrder());

    // (1) Tambah data ke kedua heap sekaligus
    public void fitur1_tambahData(int id, String nama) {
        Barang baru = new Barang(id, nama);
        minHeap.add(baru);
        maxHeap.add(baru);
    }

    // (2) Tampil Ascending
    public void fitur2_tampilkanAscending() {
        if(minHeap.isEmpty()) { System.out.println("Heap kosong!"); return; }
        PriorityQueue<Barang> temp = new PriorityQueue<>(minHeap);
        while (!temp.isEmpty()) System.out.println(temp.poll());
    }

    // (3) Tampil Descending
    public void fitur3_tampilkanDescending() {
        if(maxHeap.isEmpty()) { System.out.println("Heap kosong!"); return; }
        PriorityQueue<Barang> temp = new PriorityQueue<>(maxHeap);
        while (!temp.isEmpty()) System.out.println(temp.poll());
    }

    // (4) Hapus Root Min-Heap
    public void fitur4_hapusMin() {
        if (!minHeap.isEmpty()) {
            Barang b = minHeap.poll();
            maxHeap.remove(b); // Biar sinkron
            System.out.println("Berhasil hapus Min-Heap: " + b);
        } else System.out.println("Min-Heap kosong!");
    }

    // (5) Hapus Root Max-Heap
    public void fitur5_hapusMax() {
        if (!maxHeap.isEmpty()) {
            Barang b = maxHeap.poll();
            minHeap.remove(b); // Biar sinkron
            System.out.println("Berhasil hapus Max-Heap: " + b);
        } else System.out.println("Max-Heap kosong!");
    }

    // (6) Load data awal (Source 1, 2, 3)
    public void fitur6_loadData(String path) {
        try (Scanner sc = new Scanner(new File(path))) {
            if (sc.hasNextLine()) sc.nextLine();
            while (sc.hasNextLine()) {
                String[] p = sc.nextLine().split(",");
                if (p.length >= 2) {
                    fitur1_tambahData(Integer.parseInt(p[0].replace("\"", "").trim()), p[1].replace("\"", "").trim());
                }
            }
            System.out.println("Sistem: Data awal berhasil dimuat.");
        } catch (Exception e) {}
    }

    public static void main(String[] args) {
        heap_data_structure app = new heap_data_structure();
        app.fitur6_loadData("data100.csv");
        Scanner input = new Scanner(System.in);
        
        while (true) {
            System.out.println("\n=== MENU HEAP JAVA ===");
            System.out.println("1. Tambah Data (Manual)");
            System.out.println("2. Tampil Ascending (Min-Heap)");
            System.out.println("3. Tampil Descending (Max-Heap)");
            System.out.println("4. Hapus Min-Heap (Root)");
            System.out.println("5. Hapus Max-Heap (Root)");
            System.out.println("0. Keluar");
            System.out.print("Pilih: ");
            int pil = input.nextInt();

            if (pil == 1) {
                System.out.print("ID: "); int id = input.nextInt();
                System.out.print("Nama: "); String nm = input.next();
                app.fitur1_tambahData(id, nm);
            } else if (pil == 2) app.fitur2_tampilkanAscending();
            else if (pil == 3) app.fitur3_tampilkanDescending();
            else if (pil == 4) app.fitur4_hapusMin();
            else if (pil == 5) app.fitur5_hapusMax();
            else if (pil == 0) break;
        }
    }
}

File: heap_data_structure.py
import heapq
import pandas as pd
import os

class HeapDataStructure:
    def __init__(self):
        # Inisialisasi struktur data heap
        self.min_heap = []
        self.max_heap = []

    # (1) FITUR: Tambah data ke dalam min-heap dan max-heap sekaligus
    def fitur1_tambah_data(self, item_id, nama):
        heapq.heappush(self.min_heap, (item_id, nama))
        heapq.heappush(self.max_heap, (-item_id, nama))

    # (2) FITUR: Tampilkan urut ascending menggunakan Min-Heap
    def fitur2_tampilkan_ascending(self):
        if not self.min_heap:
            print("\n[!] Data kosong.")
            return
        print("\n--- [Fitur 2] Urutan Ascending (Min-Heap) ---")
        temp = list(self.min_heap)
        heapq.heapify(temp)
        while temp:
            id_val, nama = heapq.heappop(temp)
            print(f"ID: {id_val:<5} | Nama: {nama}")

    # (3) FITUR: Tampilkan urut descending menggunakan Max-Heap
    def fitur3_tampilkan_descending(self):
        if not self.max_heap:
            print("\n[!] Data kosong.")
            return
        print("\n--- [Fitur 3] Urutan Descending (Max-Heap) ---")
        temp = list(self.max_heap)
        heapq.heapify(temp)
        while temp:
            neg_id, nama = heapq.heappop(temp)
            print(f"ID: {-neg_id:<5} | Nama: {nama}")

    # (4) FITUR: Hapus data terkecil (Root Min-Heap)
    def fitur4_hapus_min(self):
        if self.min_heap:
            # Ambil data terkecil
            removed = heapq.heappop(self.min_heap)
            # Sinkronisasi: Hapus juga dari max_heap supaya beneran ilang
            self.max_heap = [item for item in self.max_heap if -item[0] != removed[0]]
            heapq.heapify(self.max_heap)
            print(f"\n[Fitur 4] Berhasil menghapus Root Min-Heap: {removed}")
        else:
            print("\n[!] Min-Heap sudah kosong.")

    # (5) FITUR: Hapus data terbesar (Root Max-Heap)
    def fitur5_hapus_max(self):
        if self.max_heap:
            # Ambil data terbesar
            neg_id, nama = heapq.heappop(self.max_heap)
            target_id = -neg_id
            # Sinkronisasi: Hapus juga dari min_heap supaya beneran ilang
            self.min_heap = [item for item in self.min_heap if item[0] != target_id]
            heapq.heapify(self.min_heap)
            print(f"\n[Fitur 5] Berhasil menghapus Root Max-Heap: ({target_id}, {nama})")
        else:
            print("\n[!] Max-Heap sudah kosong.")

    # (6) FITUR: Starting data awal dari file EXCEL
    def fitur6_load_excel(self, file_path):
        if os.path.exists(file_path):
            try:
                # Membaca file excel (asumsi kolom 'id' dan 'nama') [cite: 1, 2, 3]
                df = pd.read_excel(file_path)
                for _, row in df.iterrows():
                    self.fitur1_tambah_data(int(row['id']), str(row['nama']))
                print(f"\n>>> Fitur 6: Berhasil memuat {len(df)} data dari {file_path}")
            except Exception as e:
                print(f"\n[!] Error baca Excel: {e}")
        else:
            print(f"\n[!] File {file_path} tidak ditemukan.")

# --- Program Utama ---
if __name__ == "__main__":
    app = HeapDataStructure()
    
    # 6. Load data awal dari excel [cite: 1, 2, 3]
    app.fitur6_load_excel('data100.xlsx') 

    while True:
        print("\n" + "="*35)
        print("   MENU HEAP STRUCTURE (EXCEL)   ")
        print("="*35)
        print("1. Tambah Data (Manual)")
        print("2. Tampilkan Ascending (Min-Heap)")
        print("3. Tampilkan Descending (Max-Heap)")
        print("4. Hapus Root Min-Heap (Terkecil)")
        print("5. Hapus Root Max-Heap (Terbesar)")
        print("0. Keluar")
        
        pilih = input("Pilih Menu: ")
        
        if pilih == '1':
            try:
                id_input = int(input("Masukkan ID: "))
                nama_input = input("Masukkan Nama: ")
                app.fitur1_tambah_data(id_input, nama_input)
                print("Data berhasil ditambahkan ke Min & Max Heap.")
            except ValueError:
                print("ID harus angka!")
        elif pilih == '2': app.fitur2_tampilkan_ascending()
        elif pilih == '3': app.fitur3_tampilkan_descending()
        elif pilih == '4': app.fitur4_hapus_min()
        elif pilih == '5': app.fitur5_hapus_max()
        elif pilih == '0':
            print("Program selesai.")
            break
        else:
            print("Pilihan salah!")